xarray/core/dataarray.py

```diff
@@ -721,18 +726,20 @@ def __delitem__(self, key: Any) -> None:
         del self.coords[key]
 
     @property
-    def _attr_sources(self) -> List[Mapping[Hashable, Any]]:
-        """List of places to look-up items for attribute-style access"""
-        return self._item_sources + [self.attrs]
+    def _attr_sources(self) -> Iterable[Mapping[Hashable, Any]]:
+        """Places to look-up items for attribute-style access"""
+        yield from self._item_sources
+        yield self.attrs
 
     @property
-    def _item_sources(self) -> List[Mapping[Hashable, Any]]:
-        """List of places to look-up items for key-completion"""
-        return [
-            self.coords,
-            {d: self.coords[d] for d in self.dims},
-            LevelCoordinatesSource(self),
-        ]
+    def _item_sources(self) -> Iterable[Mapping[Hashable, Any]]:
+        """Places to look-up items for key-completion"""
+        yield self.coords
```

Comment:
```suggestion
        yield self._coords
```

Reply:
For iterating the keys your suggestion is definitely faster. The lookup result type would however change from `DataArray` (with `self.coords`) to `IndexVariable` (with `self._coords`) and we would lose the virtual coordinates (e.g. a dim without coordinates). 
I will change this to `yield HybridMappingProxy(keys=da._coords, mapping=da.coords)`. This improves performance of iteration without changing semantics. Thanks for the suggestion!